### PR TITLE
unittests/runtime: Migrate tests from pthreads to std::thread

### PR DESCRIPTION
**Pull request by AndrewSB**
This is required for Windows, which has std::threads, but no pthreads.
This is probably completely broken, working on running the tests locally right now.

If a member could ask @swift-ci to test, that would be much appreciated ❤️

### CODE REVIEW

**Comment by compnerd** [comment]
CC: @gparker42

**Comment by compnerd** [comment]
@swift-ci please smoke test

File: unittests/runtime/Metadata.cpp [hunk 1]
```diff
@@ -21,142 +21,80 @@
...
-    return -1;
-  }
-  barrier->count = count;
-  barrier->numThreadsWaiting = 0;
-  return 0;
-}
+  unsigned numThreads;
+  unsigned *numThreadsReady;
+  std::mutex *sharedMutex;
+  std::condition_variable *start_condition;
```
Comment (gparker42) [inline unittests/runtime/Metadata.cpp]: Can these fields be references instead of pointers?

File: unittests/runtime/Metadata.cpp [hunk 2]
```diff
@@ -249,7 +185,7 @@ TEST(Concurrent, ConcurrentList) {
   }
 
   // Check that the length of the list is correct.
-  EXPECT_EQ(ListLen, results.size() * numElem);
+  EXPECT_EQ(ListLen, 0 * numElem);
```
Comment (gparker42) [inline unittests/runtime/Metadata.cpp]: This looks wrong. Why are we verifying the contents of List and then asserting that List is empty?

**Comment by AndrewSB** [comment]
Should have said this: The patch is currently in progress, I know that's broken. I'll ping you when its ready to be reviewed.

**Comment by AndrewSB** [comment]
I think this is the patch I want. I'm testing it locally, but CI will probably go faster @compnerd / @gparker42 could you ask CI to test?

@gparker42 I might have changes to fix this, but this is more or less the change I was thinking. Took your references instead of pointer comment into account.

File: unittests/runtime/Metadata.cpp [hunk 3]
```diff
@@ -21,127 +21,75 @@
...
 }
 
 /// RaceTest(code) runs code in many threads simultaneously, 
 /// and returns a vector of all returned results.
 template <typename T, int NumThreads = 64>
-std::vector<T> 
-RaceTest(std::function<T()> code)
-{
+std::vector<T> RaceTest(std::function<T()> code) {
   const unsigned threadCount = NumThreads;
```
Comment (compnerd) [inline unittests/runtime/Metadata.cpp]: Im not sure I understand the point of the `threadCount` variable.  Why not just use `NumThreads` everywhere?

**Comment by compnerd** [comment]
@swift-ci please test

File: unittests/runtime/Metadata.cpp (hunk 3 above)
Comment (AndrewSB) [reply unittests/runtime/Metadata.cpp]: That was in the original implementation, I think it was because pthreads required an `unsigned`.
I'll remove it

**Comment by AndrewSB** [comment]
I removed `threadCount`, it looks like the CI failed with an Infra issue. Can you retry?

File: unittests/runtime/Metadata.cpp [hunk 4]
```diff
@@ -21,127 +21,73 @@
...
+		  /*code*/ code,
+		  /*result*/ nullptr,
+		  NumThreads,
+		  numThreadsReady,
+		  sharedMutex,
+		  start_condition});
+
+  // Create all the threads
+  for (unsigned i = 0; i < NumThreads; i++) {
+    threads.emplace_back(std::thread(RaceThunk<T>, contexts[i]));
```
Comment (gparker42) [inline unittests/runtime/Metadata.cpp]: This combination of sized constructor and `emplace_back()` looks wrong. I think it creates a vector of `2*NumThreads` elements, with the first half being default-initialized and the second half being the desired contents.

Comment (AndrewSB) [reply unittests/runtime/Metadata.cpp]: I'm sorry! I'm still learning cpp, my bad 😬
How should I update the vector initializer to not fill in default values? I looked at http://en.cppreference.com/w/cpp/container/vector/vector but it doesn't seem easy to have the `Allocator` not insert values. Should I switch from `emplace_back`ing to assigning to `threads[i]`?

Comment (gparker42) [reply unittests/runtime/Metadata.cpp]: For this code simply assigning to `threads[i]` should be fine. Or create an empty vector and `push_back` or `emplace_back` onto it, like the old pthreads code did. Both of these might have performance downsides, but we don't care in this code.

This stackoverflow answer has a good overview of the combinations that work and their performance effects. 
https://stackoverflow.com/a/32200517/881456

**Comment by AndrewSB** [comment]
@gparker42 that was an interesting read. Thank you 😄 
Switched to using reserving and emplacing, instead of initializing with with default-initialized

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by shahmishal** [comment]
Xcode updated

@swift-ci please test

**Comment by AndrewSB** [comment]
This looks like a legitimate failure, not just an infra issue. Investigating today

File: unittests/runtime/Metadata.cpp [hunk 5]
```diff
@@ -20,127 +20,75 @@
...
+		  NumThreads,
+		  numThreadsReady,
+		  sharedMutex,
+		  start_condition});
+
+  // Create the threads
+  std::vector<std::thread> threads;
+  threads.reserve(NumThreads);
+  for (unsigned i = 0; i < NumThreads; i++)
+    threads.emplace_back(std::bind(RaceThunk<T>, contexts[i]));
```
Comment (compnerd) [inline unittests/runtime/Metadata.cpp]: Needs a `std::cref` on the `contexts[i]`.

**Review by compnerd — CHANGES_REQUESTED** [review]

File: unittests/runtime/Metadata.cpp (hunk 5 above)
Comment (AndrewSB) [reply unittests/runtime/Metadata.cpp]: Oh right, thank you 🙏

Comment (AndrewSB) [reply unittests/runtime/Metadata.cpp]: done

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by AndrewSB** [comment]
Looks like an infra failure? 
```
12:54:42 Build step 'Execute shell' marked build as failure
12:54:42 [xUnit] [INFO] - Starting to record.
12:54:42 [xUnit] [INFO] - Processing JUnit
12:54:42 [xUnit] [INFO] - [JUnit] - 5 test report file(s) were found with the pattern 'branch-master/buildbot_incremental/swift-macosx-x86_64/swift-test-results/**/*.xml' relative to '/Users/buildnode/jenkins/workspace/swift-PR-osx' for the testing framework 'JUnit'.
12:54:42 [xUnit] [ERROR] - Test reports were found but not all of them are new. Did all the tests run?
12:54:42   * /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/swift-test-results/i386-apple-ios7.0/lit-tests.xml is 56 min old
12:54:42   * /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/swift-test-results/i386-apple-watchos2.0/lit-tests.xml is 14 min old
12:54:42   * /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/swift-test-results/x86_64-apple-ios7.0/lit-tests.xml is 42 min old
12:54:42   * /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/lit-tests.xml is 1 hr 10 min old
12:54:42   * /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/swift-test-results/x86_64-apple-tvos9.0/lit-tests.xml is 28 min old
12:54:42 
12:54:42 [xUnit] [INFO] - Failing BUILD.
12:54:42 [xUnit] [INFO] - There are errors when processing test results.
12:54:42 [xUnit] [INFO] - Skipping tests recording.
12:54:42 [xUnit] [INFO] - Stop build.
12:54:42 [BFA] Scanning build for known causes...
12:54:42 [BFA] No failure causes found
12:54:42 [BFA] Done. 0s
```

try again?

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Review by compnerd — APPROVED** [review]

**Comment by AndrewSB** [comment]
@gparker42 can you take a look last this? After your review it should be good to merge

**Pull request merged**